xdsl/parser.py

```diff
@@ -838,43 +815,38 @@ def try_parse_numerical_dims(self,
-                                       ("x", )):
-            shape = list[int](self.try_parse_numerical_dims())
-            scaling_shape: list[int] | None = None
-
-            if self.tokenizer.next_token_of_pattern("[") is not None:
-                # We now need to parse the scalable dimensions
-                scaling_shape = list(self.try_parse_numerical_dims())
-                self.parse_characters(
-                    "]", "Expected end of scalable vector dimensions here!")
-                self.parse_characters(
-                    "x", "Expected end of scalable vector dimensions here!")
+        shape = list[int](self.try_parse_numerical_dims())
+        scaling_shape: list[int] | None = None
 
-            if scaling_shape is not None:
-                # TODO: handle scaling vectors!
-                self.raise_error("Warning: scaling vectors not supported!")
-                pass
+        if self.tokenizer.next_token_of_pattern("[") is not None:
+            # We now need to parse the scalable dimensions
+            scaling_shape = list(self.try_parse_numerical_dims())
+            self.parse_characters(
+                "]", "Expected end of scalable vector dimensions here!")
+            self.parse_characters(
+                "x", "Expected end of scalable vector dimensions here!")
 
-            type = self.try_parse_type()
-            if type is None:
-                self.raise_error(
-                    "Expected a type at the end of the vector parameters!")
+        if scaling_shape is not None:
+            # TODO: handle scaling vectors!
+            self.raise_error("Warning: scaling vectors not supported!")
+            pass
+
+        type = self.try_parse_type()
+        if type is None:
+            self.raise_error(
+                "Expected a type at the end of the vector parameters!")
 
-            return VectorType.from_element_type_and_shape(type, shape)
+        return VectorType.from_element_type_and_shape(type, shape)
 
     def _parse_tensor_or_memref_dims(self) -> list[int] | None:
-        with self.tokenizer.configured(break_on=self.tokenizer.break_on +
-                                       ('x', )):
-            # Check for unranked-ness
-            if self.tokenizer.next_token_of_pattern('*') is not None:
-                # Consume `x`
-                self.parse_characters(
-                    'x',
-                    'Unranked tensors must follow format (`<*x` type `>`)')
-            else:
-                # Parse rank:
-                return list(self.try_parse_numerical_dims(lower_bound=0))
+        # Check for unranked-ness
+        if self.tokenizer.next_token_of_pattern('*') is not None:
+            # Consume `x`
+            self.parse_characters(
+                'x', 'Unranked tensors must follow format (`<*x` type `>`)')
+        else:
+            # Parse rank:
+            return list(self.try_parse_numerical_dims(lower_bound=0))
```

Comment:
How good is our test coverage over the parsing of these attributes? I am a little worried about these refactorings (if we don't have test coverage that is). If there's coverage then I'm fine!

Reply:
https://github.com/xdslproject/xdsl/blob/main/tests/filecheck/mlir-conversion/builtin_attrs.mlir should test them.
And coverage seems okay: https://app.codecov.io/gh/xdslproject/xdsl/blob/main/xdsl/parser.py?trend=3%20months#L839

But more importantly, how the hell did you manage to comment on so many lines? I thought you could only comment on a single line?!

Reply:
Okay, I just saw how to do that, that's really really cool!